icon.js

```diff
@@ -18,31 +18,38 @@ var PX_HEIGHT = 32
 
 var sheet_hash
 
+var images_buffer = new Map()
+
 function Sprite(name, col, row) {
     this.name = name
     this.icon_col = col
     this.icon_row = row
-}
+}   
 
 function getImage(obj, suppressTooltip, tooltipTarget) {
-    var im = blankImage()
-    im.classList.add("icon")
-    var x = -obj.icon_col * PX_WIDTH
-    var y = -obj.icon_row * PX_HEIGHT
-    im.style.setProperty("background", "url(images/sprite-sheet-" + sheet_hash + ".png)")
-    im.style.setProperty("background-position", x + "px " + y + "px")
-    if (tooltipsEnabled && obj.renderTooltip && !suppressTooltip) {
-        addTooltip(im, obj, tooltipTarget)
+    if (images_buffer.has(obj.name)) {
+        return images_buffer.get(obj.name)
     } else {
-        im.title = obj.name
+        var im = blankImage()
+        im.classList.add("icon")
+        var x = -obj.icon_col * PX_WIDTH
+        var y = -obj.icon_row * PX_HEIGHT
+        im.style.setProperty("background", "url(images/sprite-sheet-" + sheet_hash + ".png)")
+        im.style.setProperty("background-position", x + "px " + y + "px")
+        if (tooltipsEnabled && obj.renderTooltip && !suppressTooltip) {
+            addTooltip(im, obj, tooltipTarget)
+        } else {
+            im.title = obj.name
+        }
+        im.alt = obj.name
+        images_buffer[obj.name] = im
```

Comment:
There are several problems with the image caching approach.
* The correct way to add items to a Map is be [Map.prototype.set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/set)
* Caching images caches tooltips shown on them, which would quickly result in incorrect information.
* Different things with the same name are allowed to have different icons. For example, `crude-oil` the icon has a different image than `crude-oil` the resource. This is the only difference I know of in base data, but the issue could be more prevalent in modded games.
* It is not possible to have the same DOM node in a document twice. See the [appendChild documentation](https://developer.mozilla.org/en-US/docs/Web/API/Node/appendChild)

Reply:
thanks for the review.

Reply:
I should've noted that the first commit speeds up `plusHandler` from 270ms to 170ms on my machine, and looks like a good idea. I'd also like to note that I can't speak for KirkMcDonald, I'm just a drive-by reviewer.